### PR TITLE
[Data-Plane Mgr] Add Local Cache Support to DPM

### PR DESCRIPTION
**Pull request by chenpiaoping**
1. When creating a port, save the basic information of all port in all subnet to the local cache of dpm. 
2. When creating router configuration get the basic information of port from the local cache of dpm according to subnet id.
3. Rename DataPlane to Dpm: DataPlaneController--->DpmController, DataPlaneService--->DpmService.
4. Split DpmService into multiple files according to resource type.
5. Modify the content of dpm processing result.

### CODE REVIEW

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/DataPlaneClient.java [hunk 1]
```diff
@@ -26,15 +26,7 @@
 
 @Component
 public interface DataPlaneClient {
-    Map<String, List<GoalStateOperationStatus>> createGoalStates(GoalState goalState, String hostIp) throws Exception;
-
-    List<Map<String, List<GoalStateOperationStatus>>> createGoalStates(List<UnicastGoalState> unicastGoalStates) throws Exception;
-
-    List<Map<String, List<GoalStateOperationStatus>>> updateGoalStates(List<UnicastGoalState> unicastGoalStates) throws Exception;
-
-    List<Map<String, List<GoalStateOperationStatus>>> deleteGoalStates(List<UnicastGoalState> unicastGoalStates) throws Exception;
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/DataPlaneClient.java:35]: A quick question: why do we remove these interfaces?

Comment (xieus) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/DataPlaneClient.java:35]: Actually it is okay to keep the interfaces which reminds us that we need to implement in the future.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/ovs/ResourceService.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        distributed under the License is distributed on an "AS IS" BASIS,
+        WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+        See the License for the specific language governing permissions and
+        limitations under the License.
+*/
+package com.futurewei.alcor.dataplane.service.ovs;
+
+import com.futurewei.alcor.schema.Common.OperationType;
+
+public class ResourceService {
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/ovs/ResourceService.java:20]: ResourceService class could be moved on layer up to com.futurewei.alcor.dataplane.service;

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        List<InternalPortEntity> portEntities = networkConfig.getPortEntities();
+        if (portEntities == null) {
+            return;
+        }
+
+        Map<String, InternalSubnetPorts> subnetPortsMap = new HashMap<>();
+        for (InternalPortEntity portEntity: portEntities) {
+            List<PortEntity.FixedIp> fixedIps = portEntity.getFixedIps();
+            if (fixedIps == null) {
+                continue;
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java:50]: Suggest to log this unusual portEntity (at least on the DEBUG level) as this is an unexpected scenario.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+                InternalSubnetPorts subnetPorts = subnetPortsMap.get(subnetId);
+                if (subnetPorts == null) {
+                    SubnetEntity subnetEntity = getSubnetEntity(networkConfig, fixedIp.getSubnetId());
+                    subnetPorts = new InternalSubnetPorts();
+                    subnetPorts.setSubnetId(subnetId);
+                    subnetPorts.setGatewayPortMac(subnetEntity.getGatewayMacAddress());
+                    subnetPorts.setGatewayPortIp(subnetEntity.getGatewayIp());
+
+                    //FIXME: get the gateway port id
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java:71]: GW port could be retrieved from SubnetEntity I believe.

**Review by xieus — COMMENTED** [review]
@chenpiaoping Do we have the logics to use the cache in case of cache hit and cache miss?

Left a few minor comments. The rest looks pretty good to me.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/DataPlaneClient.java (hunk 1 above)
Comment (chenpiaoping) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/DataPlaneClient.java:35]: I was thinking about adding it when I needed it.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java (hunk 3 above)
Comment (chenpiaoping) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java:50]: Good idea

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java (hunk 4 above)
Comment (chenpiaoping) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java:71]: OK, I'll check it out.

**Comment by chenpiaoping** [comment]
> @chenpiaoping Do we have the logics to use the cache in case of cache hit and cache miss?
Not yet.

**Comment by xieus** [comment]
@chenpiaoping Let us get back on this PR. It has been a while.

**Comment by xieus** [comment]
BTW https://github.com/futurewei-cloud/alcor/pull/468 is merged to master. After the cache is in, we can start testing E2E.

**Comment by chenpiaoping** [comment]
> @chenpiaoping Let us get back on this PR. It has been a while.

ok

**Comment by chenpiaoping** [comment]
> > @chenpiaoping Do we have the logics to use the cache in case of cache hit and cache miss?
> > Not yet.

Under what circumstances will miss occur?

**Comment by xieus** [comment]
> > @chenpiaoping Let us get back on this PR. It has been a while.
> 
> OK,there seems to be a lot of conflict.

yeah, DPM is very popular these days :-)

**Comment by xieus** [comment]
> > > @chenpiaoping Do we have the logics to use the cache in case of cache hit and cache miss?
> > > Not yet.
> 
> Under what circumstances will miss occur?

In an ideal situation, the port creation should be always reach DPM earlier than RM reaches DPM for routing rule update. We can image that in some rare (but possible) race condition, the second call could reach DPM earlier than the first call.

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java (hunk 4 above)
Comment (chenpiaoping) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java]: fixed

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java (hunk 3 above)
Comment (chenpiaoping) [reply services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/cache/LocalCacheImpl.java:56]: fixed

File: services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/DataPlaneClientImpl.java [hunk 5]
```diff
@@ -102,22 +95,32 @@ private void createGoalState(MulticastGoalState multicastGoalState) throws Excep
...
+
+        return failedHosts;
     }
 
     @Override
-    public List<Map<String, List<Goalstateprovisioner.GoalStateOperationReply.GoalStateOperationStatus>>> createGoalStates(List<UnicastGoalState> unicastGoalStates) throws Exception {
+    public List<String> createGoalStates(List<UnicastGoalState> unicastGoalStates) throws Exception {
+        List<String> failedHosts = new ArrayList<>();
+
         for (UnicastGoalState unicastGoalState: unicastGoalStates) {
```
Comment (xieus) [inline services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/DataPlaneClientImpl.java:124]: In the case of large number of neighbors, this for-loop could slow down potentially. See if we can optimize this part. @chenpiaoping

**Review by xieus — APPROVED** [review]
Overall LGTM. Left a comment on one place that might have potential issue. Please take a look to see if we could further optimize shortly. @chenpiaoping

**Pull request merged**